project-koku/koku
# Cost 2639 targeted provider delete cascade

Fixes [COST-2639](https://issues.redhat.com/browse/COST-2639)

Changes proposed in this PR:
* Create new targeted-SQL cascade_delete function
    * This is using catalog SQL queries to get linked table information
    * The code also uses name matching to get top-level tables and specific `api_*` tables
    * The code will also follow 2nd-level linking as well

Testing Instructions:
1. Checkout this branch
2. `make docker-reinitdb-with-sources`
3. `make load-test-customer-data`
4. `python ./koku/manage.py shell`
5. Execute the following comands
    ```
    from koku.database import get_model
    Provider = get_model("Provider")
    paws = Provider.objects.filter(type__contains="AWS").first()
    paws is not None  # Verify that we got a record
    paws.delete()  # You will see logs for each action taken and number of recs deleted
    Provider.objects.filter(uuid=paws.pk).count()  # Should return zero
    #
    # Try it with other provider types as well
    #
    ```
